src/RevProc.cc

```diff
@@ -1530,43 +1530,21 @@ void RevProc::HandleALUFault(unsigned width){
 }
 
 bool RevProc::DependencyCheck(unsigned HartID, const RevInst* I) const {
-
-  const auto* E = &InstTable[I->entry];
-  const auto* regFile = GetRegFile(HartID);
-
-  // check LS queue for outstanding load - ignore r0
-  for(const auto& [reg, regClass] : {
-      std::tie(I->rs1, E->rs1Class),
-      std::tie(I->rs2, E->rs2Class),
-      std::tie(I->rs3, E->rs3Class),
-      std::tie(I->rd,  E->rdClass) }){
-    if((reg != 0) && (regFile->GetLSQueue()->count(make_lsq_hash(reg,
-                                                                 regClass,
-                                                                 HartID))) > 0){
-      return true;
-    }
-  }
-
-  // Iterate through the source registers rs1, rs2, rs3 and find any dependency
-  // based on the class of the source register and the associated scoreboard
-  for(const auto& [reg, regClass] : {
-      std::tie(I->rs1, E->rs1Class),
-      std::tie(I->rs2, E->rs2Class),
-      std::tie(I->rs3, E->rs3Class) }){
-    if(reg < _REV_NUM_REGS_){
-      switch(regClass){
-      case RevRegClass::RegFLOAT:
-        if(regFile->FP_Scoreboard[reg]) return true;
-        break;
-      case RevRegClass::RegGPR:
-        if(regFile->RV_Scoreboard[reg]) return true;
-        break;
-      default:
-        break;
-      }
-    }
-  }
-  return false;
+  const RevRegFile* regFile = GetRegFile(HartID);
+  const RevInstEntry* E = &InstTable[I->entry];
+
+  return
+    // check LS queue for outstanding load
+    LSQCheck(HartID, regFile, I->rs1, E->rs1Class) ||
+    LSQCheck(HartID, regFile, I->rs2, E->rs2Class) ||
+    LSQCheck(HartID, regFile, I->rs3, E->rs3Class) ||
+    LSQCheck(HartID, regFile, I->rd , E->rdClass) ||
+
```

Comment:
we were checking `rd` previously, but not sure that is necessary in all cases? We may be leaning on it for atomics in an unexpected way, but we should experiment with removing in as in the general case a WAW hazard is non blocking

Reply:
WAW could be a problem with atomics with a threaded simulation or any out-of-order processing (do we atomically increment or decrement first?).